cpp/arcticdb/storage/storage.hpp

```diff
@@ -127,39 +132,58 @@ class Storage : public StorageBase {
     }
 
     void remove(Composite<VariantKey> &&ks, RemoveOpts opts) {
-        derived().do_remove(std::move(ks), opts);
+        do_remove(std::move(ks), opts);
     }
 
     void remove(VariantKey&& key, RemoveOpts opts) {
         return remove(Composite<VariantKey>{std::move(key)}, opts);
     }
 
     bool supports_prefix_matching() {
-        return derived().do_supports_prefix_matching();
+        return do_supports_prefix_matching();
     }
 
     bool fast_delete() {
-        return derived().do_fast_delete();
+        return do_fast_delete();
     }
 
     inline bool key_exists(const VariantKey &key) {
-        return derived().do_key_exists(key);
+        return do_key_exists(key);
     }
 
     void iterate_type(KeyType key_type, const IterateTypeVisitor& visitor, const std::string &prefix = std::string()) {
-        derived().do_iterate_type(key_type, visitor, prefix);
+        do_iterate_type(key_type, visitor, prefix);
+    }
+
+    std::string storage_specific(const VariantKey& key) {
+        return do_storage_specific(key);
     }
 
     [[nodiscard]] const LibraryPath &library_path() const { return lib_path_; }
     [[nodiscard]] OpenMode open_mode() const { return mode_; }
 
+protected:
+    virtual void do_write(Composite<KeySegmentPair>&& kvs) = 0;
+
+    virtual void do_update(Composite<KeySegmentPair>&& kvs, UpdateOpts opts) = 0;
+
+    virtual void do_read(Composite<VariantKey>&& ks, const ReadVisitor& visitor, ReadKeyOpts opts) = 0;
+
+    virtual void do_remove(Composite<VariantKey>&& ks, RemoveOpts opts) = 0;
+
+    virtual bool do_key_exists(const VariantKey& key) = 0;
+
+    virtual bool do_supports_prefix_matching() = 0;
+
+    virtual bool do_fast_delete() = 0;
+
+    virtual void do_iterate_type(KeyType key_type, const IterateTypeVisitor& visitor, const std::string & prefix) = 0;
+
+    virtual std::string do_storage_specific(const VariantKey& key) = 0;
```

Comment:
Note for reviewers: as part of this PR, should we `const`-qualified methods which can? If so, should we also `const`-qualify their associated public method?

Reply:
Makes sense to me, is there a reason not to that you can think of?

Reply:
I do not think of a reason not to use `const`-qualification when we actually can.

Reply:
Yeh makes sense, no need to overcomplicate it right now